tools/devtool

```diff
@@ -375,6 +408,117 @@ run_devctr() {
+    declare allowed_ranges="$2"
+    shift
+    shift
+    options="$@"
+    options+=('Skip')
+    declare selection
+    select opt in ${options[@]}; do
+        if [[ $opt = "Skip" ]]; then
+            die "Skipping selection of package ${package_name}"
+        elif [ "$REPLY" -ge 1 ] && [ "$REPLY" -le $allowed_ranges ]; then    
+           echo "$opt"
+           break
+        else
+           say_err "Invalid selection "$REPLY", try selecting a valid option."
+        fi
+    done
+}
+
+# Runs a curl on the Pypi package store and returns all the
+# available version of the package sorted in ascending order
+# of their version using version sort.
+# TL;DR - "jq" should be installed on local host for this function to work.
+#
+# Args:
+#   $1  python package name.
+#  
+get_all_availiable_versions() {
+    declare package_name="$1"
+    PACKAGE_JSON_URL="https://pypi.org/pypi/${package_name}/json"
+    curl -L -s "$PACKAGE_JSON_URL" | jq -r '.releases | keys | .[]' | sort -V
+}
+
+# Looks for any package which is defined/added recently in
+# requirements.txt file but not present in {arch-based}-requirements-freeze.txt file.
+# For every diff it prompts the user to pick from top k(set to 5) available versions
+# and updates the freezed requirements file with new freezed pacage version.
+# 
+# Args:
+#   $1  path to the {arch-based}-requirements-freeze file.
+# 
+update_packages() {
+    declare requirements_freeze_file_location="$1"  
+    declare parsed_package_location=$(mktemp)
+    say "Checking for newly added requirements..."
+    while IFS= read -r line; do
+        echo "${line%%=*}" >>$parsed_package_location
+    done <"$requirements_freeze_file_location"
+    unset IFS
+
+    diffs=($(grep -v -f $parsed_package_location $FC_DEVCTR_DIR/requirements/requirements.txt))    
+    rm $parsed_package_location
+
+    for package_name in "${diffs[@]}"; do
+        # get top 5 available versions
+        top_k_versions=($(get_all_availiable_versions $package_name | tail -n 5))
+        echo "${YELLOW}Newer package requirement ${RED}${package_name}${RESET} detected ${RESET}"
+        selected_version=($(select_python_package_version $package_name ${#top_k_versions[@]} ${top_k_versions[@]}))
+
+        if [ ! -z "$selected_version" ]; then
+            echo "${GREEN}Freezing package ${RED}${package_name}${RESET} with version ${RED}${selected_version} ${RESET}"
+            freezed_version=${package_name}==${selected_version}
+            # add new requirement with freezed version.
+            echo $freezed_version >>$requirements_freeze_file_location
+        fi
+    done
+}
+
+# Scans {arch-based}-requirements-freeze.txt file and looks for all package versions greater
+# than current freezed versions. Prompts the user to either upgrade the
+# {arch-based}-requirements-freeze.txt file with any of the newer versions or skip the upgrade.
+# 
+# Args:
+#   $1  path to the {arch-based}-requirements-freeze file.
+# 
+upgrade_version() {
```

Comment:
pip provides a subcommand for checking for outdated packages: `pip list --outdated`

It would be much simpler to run this command and check if there are outdated packages. If there are, we just prompt the user and say something like: 
````
There are outdated packages: 
<output of pip list --outdated>. 

Would you like to upgrade these packages?
```

Reply:
Agree that the current logic is little complex. One reason I didn't choose ```pip list --outdated``` was because, on the host, there might be user installed packages other than defined in requirements-freeze.txt. If we run pip list --outdated and then freeze it back we bring unwanted changes and its will become hard to filter/identify what to pick and leave from git diff.

Had ```pip list --outdated``` taken a requirements files, that would have solved our problem.